proof/refine/ARM/Invariants_H.thy

```diff
@@ -1616,26 +1616,28 @@ lemma ps_clear_def2:
   apply (drule plus_one_helper, simp)
   done
 
+lemmas in_omonad = ofail_def oreturn_def oassert_def oassert_opt_def asks_def
```

Comment:
There is an existing `in_omonad` in `lib/Monad_WP/OptionMonadND` -- please extend there instead of shadowing it, because it's a set used in a lot of places (AutoCorres, AInvs etc).

But: I don't think these definitions should be just unfolded -- `in_monad` and `in_omonad` are for the situation `(s', r) : monad s` or `monad s = Some r`. Only the rules that take apart these constructs into smaller components should be in that set.

We could make a set `omonad_defs` that contains the definitions you have here. That would then be use to always unfold all primitive monad operators. Then again, we should not need to do that often apart from when we're developing the monad framework (i.e. the equivalent of wp rules, which is what the `in_omonad` set is largely supposed to accomplish).

Still fine to define that `omonad_defs` set, we should just be aware that using it is a sign that we're not done with the framework yet. The only thing that is a bit annoying is `ohaskell_fail_def` and `ohaskell_assert_def`, which both should be defined only in `Haskell_Lib`, not in `OptionMonad*`. It'd be Ok to extend the collection there, though.

Reply:
In `Haskell_Lib`, `lookup` is not available, while it seems to fine to put `ohaskell_fail` and the other in `OptionMonad` along with others (they just take one extra argument that is discarded). Are you saying they shouldn’t be there because it is about haskell, or some other reasons?

I’ll remove `in_omonad` from here and add `omonad_defs` (in OptionMonad maybe).

Reply:
> Are you saying they shouldn’t be there because they are about haskell, or some other reasons?

Yes. Since we're now using the option monad in Haskell, it'd be perfectly fine to include those monads there already, i.e. add the `OptionMonad` theory you need to the `import` statements there. This would also make it naturally available to the rest of `Refine`.